janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -414,63 +423,50 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
                 .getOrElseThrow(EXCEPTION_MAPPER);
     }
 
     /**
      * This class provides a paging implementation that sits on top of the DSE Cassandra driver. The driver already
      * has its own built in paging support but this has limitations when doing a full scan of the key ring due
      * to how driver paging metadata is stored. The driver stores a full history of a given query's paging metadata
-     * which can lead to OOM issues on non-trivially sized data sets. This class overcomes this by doing another level
-     * of paging that re-executes the query after a configurable number of rows. When the original query is re-executed
-     * it is initialized to the correct offset using the last page's metadata.
+     * which can lead to OOM issues on non-trivially sized data sets. This class overcomes this by forcing the internal
+     * metadata which isn't needed anymore to be deleted to be free for GC.
      */
     private class CQLPagingIterator implements Iterator<Row> {
 
-        private ResultSet currentResultSet;
-
-        private int index;
-        private int paginatedResultSize;
-        private final Supplier<Statement> statementSupplier;
-
-        private byte[] lastPagingState = null;
+        private final List<ExecutionInfo> dseStoredExecutionInfos;
+        private final java.util.Iterator<Row> currentPageIterator;
 
-        public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSupplier) {
-            this.index = 0;
-            this.paginatedResultSize = pageSize;
-            this.statementSupplier = statementSupplier;
-            this.currentResultSet = getResultSet();
+        public CQLPagingIterator(BoundStatement boundStatement) {
+            ResultSet currentResultSet = session.execute(boundStatement);
+            currentPageIterator = currentResultSet.iterator();
+            this.dseStoredExecutionInfos = currentResultSet.getExecutionInfos();
         }
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            return currentPageIterator.hasNext();
         }
 
         @Override
         public Row next() {
-            if(index == paginatedResultSize) {
-                currentResultSet = getResultSet();
-                this.index = 0;
-            }
-            this.index++;
-            lastPagingState = currentResultSet.getExecutionInfo().getPagingStateUnsafe();
-            return currentResultSet.one();
-
-        }
-
-        private ResultSet getResultSet() {
-            final Statement boundStmnt = statementSupplier.get();
-            if (lastPagingState != null) {
-                boundStmnt.setPagingStateUnsafe(lastPagingState);
+            if(dseStoredExecutionInfos.size()>1){
+                if(dseStoredExecutionInfos.size()==2){
+                    dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1));
```

Comment:
What about `dseStoredExecutionInfos.remove(0)`? Actually, if we simply do `dseStoredExecutionInfos.remove(0)`, or use your code in the `else` branch, then we don't need two branches at all. Is there any specific reason to keep the `if(dseStoredExecutionInfos.size()==2)` condition here?

Reply:
`dseStoredExecutionInfos.remove(0)` - will remove only the first element. So, it there are 3 elements 2 elements will be left. So, in case we execute `next(); hasNext(); next(); hasNext(); next(); hasNext();` with page size `1` it will produce `5` elements in `dseStoredExecutionInfos` because on `next(); hasNext();` we are adding 2 elments but on `nest()` we are removing only 1 element.
`dseStoredExecutionInfos.remove(0)` and `dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1))` are kind of the same if the `size == 2` but when we execute `dseStoredExecutionInfos.remove(1)` we are not triggering `System.arraycopy` because we removed the last element and when we execute `dseStoredExecutionInfos.remove(0)` we trigger that. From another hand `System.arraycopy` will copy just 1 element, thus it will be fast. I just thought that `dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1))` would be several operations faster then `dseStoredExecutionInfos.remove(0)` but it shouldn't be much difference and I didn't really compare the performance. Just assumed

Reply:
Ah, I got it. I misunderstood second part. You are right. We don't need 2 branches if we use the code in `else` branch. I was just optimizing usage of unnecessary operations here but I think it just made the code more complex

Reply:
Updated the code as you suggested here: https://github.com/JanusGraph/janusgraph/compare/70321725c0cd1be7a40ceedd36a43a94127d0eb1..65cad01828f1a455cc51435f8efaf309111c0768